lib/galaxy/datatypes/data.py

```diff
@@ -1212,6 +1212,18 @@ def regex_line_dataprovider(
 class Directory(Data):
     """Class representing a directory of files."""
 
+    file_ext = "directory"
+
+    def _archive_main_file(
```

Comment:
Can you add a test for roundtripping a directory via the API ?

Reply:
Sure I can try!

Reply:
I've added a test, but I'm not sure why I'm getting `ModuleNotFoundError: No module named 'galaxy'` when running the converter tool... using the UI seems to work fine :thinking:

Reply:
Ahh I see, so you mean something like this https://github.com/galaxyproject/galaxy/commit/fc7e959e8fc1a271131759ea26fc18a5f3ced6ad#diff-a6ab1700bcef9e1585a2bb0f84e8888470a770fb81c3e0337930e7cad573093fR662

I'll do that :+1:

Reply:
I've tried this:

```python
    def test_fetch_directory(self, history_id):
        testdir = TestDataResolver().get_filename("testdir.tar")
        with open(testdir, "rb") as fh:
            details = self._upload_and_get_details(
                fh, api="fetch", history_id=history_id, ext="directory", assert_ok=True
            )
        assert details["file_ext"] == "directory"
        assert details["file_size"] == 3584
        content = self.dataset_populator.get_history_dataset_content(
            history_id, dataset=details, to_ext="directory", type="bytes"
        )
        dir_path = decompress_bytes_to_directory(cast(bytes, content))
        assert dir_path.endswith("testdir")
        for path, entry_class in EXPECTED_CONTENTS.items():
            path = os.path.join(dir_path, os.path.pardir, path)
            if entry_class == "Directory":
                assert os.path.isdir(path)
            else:
                assert os.path.isfile(path)
```

But if I don't run the converter manually instead of the `to_ext="directory"` the extra_files_path is empty, I guess that is why you have more changes persisting extra files in the object store in your referenced branch https://github.com/galaxyproject/galaxy/compare/dev...mvdbeek:galaxy:directory_datatype_improvements#diff-8640d91ef47bca302b00039012979f4b1b79f5dbffbe2431bc9a05f19fb4c7d0R132

Should we merge your branch instead? Is something still missing in your branch or should that be how to do it?

Sorry, I'm a bit lost :sweat_smile:

Reply:
@mvdbeek, re-reading your comment:

> In either case if you upload a tar file and download it again that should be sufficient.

 do you mean something simpler like this instead?

```python
def test_upload_tar_roundtrip(self, history_id):
        testdir = TestDataResolver().get_filename("testdir.tar")
        expected_hash = md5_hash_file(testdir)
        expected_size = os.path.getsize(testdir)
        with open(testdir, "rb") as fh:
            details = self._upload_and_get_details(fh, api="fetch", history_id=history_id, assert_ok=True)
        assert details["file_ext"] == "tar"
        assert details["file_size"] == expected_size
        content = cast(
            bytes, self.dataset_populator.get_history_dataset_content(history_id, dataset=details, type="bytes")
        )
        assert len(content) == expected_size
        dir_path = decompress_bytes_to_directory(content)
        expected_contents = {
            "testdir": "Directory",
            "testdir/c": "Directory",
            "testdir/a": "File",
            "testdir/b": "File",
            "testdir/c/d": "File",
        }
        assert dir_path.endswith("testdir")
        for path, entry_class in expected_contents.items():
            path = os.path.join(dir_path, os.path.pardir, path)
            if entry_class == "Directory":
                assert os.path.isdir(path)
            else:
                assert os.path.isfile(path)

        with tempfile.NamedTemporaryFile("wb") as temp:
            temp.write(content)
            actual_hash = md5_hash_file(temp.name)
            assert actual_hash == expected_hash
```

If this is what you mean, the uploaded vs downloaded tar size and contents match, but the hashes don't (not sure why). 
I still don't see the connection with the `directory` datatype in this case if we are not using the converter, so I might be misunderstanding something :disappointed:

Reply:
You've overwritten `_archive_main_file`, as a result you've made sure you're not getting an (empty) extra file added to the archive.

`test_upload_tar_roundtrip` looks fine to me, you sure you don't need to flush and that that is why the checksums don't match ? You could also skip the hash and just compare the bytes. If the contents are the same, no added or removed files or structure then it's all good.

Reply:
Ahhh that was it... I was missing the flush :see_no_evil: 

Thank you very much!!